### PR TITLE
More options for styling and changes to boxplot

### PR DESCRIPTION
**Pull request by sethaxen**
Note: depends on https://github.com/JuliaPlots/AbstractPlotting.jl/pull/367

This PR primarily adds a lot more style options to the recipes and substantially refactors `BoxPlot`. A summary of the changes:
- `Violin` now internally calls `poly!` instead of `mesh!`
```julia
violin(Data(mtcars), :Cyl, :MPG)
```
Before:
violin_before
After (introduces some aliasing from `poly`):
violin_after

```julia
violin(Data(mtcars), :Cyl, :MPG; color = :orange, strokecolor = :black, strokewidth = 2, show_median = true)
```
Before:
violin_color_before
After (the misalignment is not new, it's just hidden with the white stroke):
violin_color_after

- `BoxPlot`'s box and notch code has been refactored into a `CrossBar` recipe. In addition to being cleaner, this will make it easier to create legends for boxplots that contain small boxplots with the same styling. `CrossBar` now internally creates a mesh, which circumvents errors when the notches exceed the IQR.
```julia
crossbar(1:5, 1:5, 0.5:0.5:2.5, 2:2:10)
```
crossbar
```julia
crossbar(1:5, 1:5, 0.5:0.5:2.5, 2:2:10; notch = true, notchmin = 0.75:0.75:3.75, notchmax = 1.5:1.5:7.5)
```
crossbar_notch

- Setting `whisker_width` now works (before it was unused). It now acts as a multiplier of `width`.
- Fixed computation of notch values (should be `median +/- nw`, but was `(q2 + nw / 2, q4 - nw / 2)`, where `nw` is the output of `notch_width`).
```julia
boxplot(Data(mtcars), zero(mtcars.MPG), :MPG; notch = true, whisker_width=0.1)
```
Before:
boxplot_whiskers_before
After:
boxplot_whiskers_after
```julia
boxplot(Data(ToothGrowth), :Dose, :Len; notch = true, markersize = 10px)
```
Before:
boxplot_notch_before
After:
boxplot_notch_after
vs ggplot2:
Rplot

- `ErrorBar` and its variants can now show the T whisker styles
```julia
errorbary(1:5, 1:5, 0.1:0.1:0.5; whisker_width = 0.05:0.05:0.25)
```
Before:
errorbary_whiskers_before
After:
errorbary_whiskers_after
```julia
errorbar(1:5, 1:5, 0.1:0.1:0.5, 0.1:0.1:0.5; xwhisker_width = 0.05:0.05:0.25, ywhisker_width = 0.1:0.1:0.5)
```
Before:
errorbar_whiskers_before
After:
errorbar_whiskers_after

While at it, I had a few stylistic recommendations to discuss:
1. `ErrorBar` defaults to no T's, while `BoxPlot` defaults to T's for its whiskers. I think that no T's makes more sense, as T's in the boxplot add no new information.
2. I also think  it makes more sense for `ErrorBar` and its variants to take signatures like `(x, y, xmin, xmax, ymin, ymax)` for more customizability. One could always provide a `convert_arguments` method that takes `(x, y, Δx, Δy) -> (x, y, x-Δx, x+Δx, y-Δy, y+Δy)` so that the same signatures are covered. This would also e.g. let `BoxPlot` use `ErrorBar` to draw its whiskers.

### CODE REVIEW

**Comment by mkborregaard** [comment]
I always feel it would be great with "before" and "after" plots + syntax examples in the comment when making PRs on plotting packages. Given that it's so visual. Would that be reasonably easy to add?

**Comment by asinghvi17** [comment]
I guess we kind of already have that with refimages?  All it would take is a plot with theming v/s the refimage, which we can always add more of

**Comment by mkborregaard** [comment]
Yes exactly I just meant like https://github.com/JuliaPlots/Plots.jl/pull/1561#issuecomment-398652772

**Comment by sethaxen** [comment]
@mkborregaard Alright, I updated the original post with images to reflect the main changes. There are too many new style options to demo them all.

I should note that the crossbar code is still a little bit finnicky when the notches exceed the box dimensions e.g. this example breaks in this PR but is fine before this PR (except for wrong notches):

```julia
boxplot(Data(mpg), :Class, :Hwy; notch = true, markersize = 10px)
```
boxplot_notch_byebye

Here's another example where notches exceed the box limits:

```julia
boxplot(Data(mtcars), :VS, :MPG; notch = true, markersize = 10px)
```
boxplot_notch2

I'd prefer how ggplot2 shows it, where there's some visual indication that something is wrong

```r
ggplot(mtcars, aes(y=mpg, x=factor(vs))) + geom_boxplot(notch = TRUE)
```
Rplot01

Does anyone know how I can do that with a mesh or what could be causing the box to go invisible above?

**Comment by mkborregaard** [comment]
Wow nice

**Comment by piever** [comment]
Really nice work!

A few scattered comments:

- (Possibly unrelated to this PR). I've noticed that `BoxPlot` centers its boxes at `0.5, 1, 2`, which I imagine are the possible categorical values. Should they be equally spaced, even if it is a numeric axis?

- The reason we chose `mesh!` instead of `poly!` at the time was that otherwise it didn't work well with colors. Have you checked that say `violin(Data(mtcars), :Cyl, :MPG, color = [:red, :blue, :green])` does the right thing (colors the three violins differently)?

- I like no T for errors as a default. I confess I don't really understand boxplots, what information do the Ts add there? Otherwise maybe also no T for boxplot could be a default (from what I understand the vertical line reaches the maximum so the information is preserved). A more extreme view is to draw boxplots as in the image below. The dot is the median and the vertical lines go from minimum to interquartile range, so same information but less ink.  

- I confess I'm not 100% happy of signatures like `(x, y, xmin, xmax, ymin, ymax)`. I think six arguments is too many and it is hard to remember the order. We have an interval type `a..b`, so maybe we could allow `Δx` to be an interval, say `-0.2..0.1` to have an asymmetric error bar around. In the method we could simply convert `Δx::Real` to `-Δx...Δx` at the beginning.

**Comment by sethaxen** [comment]
> * (Possibly unrelated to this PR). I've noticed that `BoxPlot` centers its boxes at `0.5, 1, 2`, which I imagine are the possible categorical values. Should they be equally spaced, even if it is a numeric axis?

I think more useful would be some utility for marking a "column" to be treated as categorical. Basically something like ggplot2's `factor`. But it might be better to address this in the AlgebraOfGraphics package, because there are some related things that would be useful. e.g. wanting to reverse/reorder the factors is common (and really inconvenient in R) or discretize a continuous variable (e.g. to make boolean or divide into intervals). A syntax for this kind of thing would be nice.

> * The reason we chose `mesh!` instead of `poly!` at the time was that otherwise it didn't work well with colors. Have you checked that say `violin(Data(mtcars), :Cyl, :MPG, color = [:red, :blue, :green])` does the right thing (colors the three violins differently)?

Yep, works just fine.
violin_colors

> * I like no T for errors as a default. I confess I don't really understand boxplots, what information do the Ts add there? Otherwise maybe also no T for boxplot could be a default (from what I understand the vertical line reaches the maximum so the information is preserved).

The Ts show no new information. I think their inclusion is mostly to make the end more visible, but I don't think they really do that. The version without T's is also widely used, and I'd prefer that as the default.

> A more extreme view is to draw boxplots as in the image below. The dot is the median and the vertical lines go from minimum to interquartile range, so same information but less ink. 

Yeah, this is (Tufte's) midgap plot. [This paper](https://vita.had.co.nz/papers/boxplots.pdf) by Hadley Wickham references a paper that indicates it causes perceptual inaccuracies. While it doesn't reduce the ink-to-data ratio, I do like the `geom_pointinterval` in tidybayes (below) as a compromise that retains the simplicity.
slabinterval_family (2)

> * I confess I'm not 100% happy of signatures like `(x, y, xmin, xmax, ymin, ymax)`. I think six arguments is too many and it is hard to remember the order. We have an interval type `a..b`, so maybe we could allow `Δx` to be an interval, say `-0.2..0.1` to have an asymmetric error bar around. In the method we could simply convert `Δx::Real` to `-Δx...Δx` at the beginning.

Yeah, I agree it's not nice. I realized that if  using `xmin` and `xmax`, then no `x` is needed.
There are a few visualizations that just show a range with maybe a centrality parameter (see [these examples](https://ggplot2.tidyverse.org/reference/geom_linerange.html).
It would be great to have some more of these but with compatible signatures. Perhaps `crossbar` could have the signatures `(x, y, ymin..ymax)`,`(x, ymin..ymax)`(this one hides the midline), `errorbary` could have the signatures `(x, ymin..ymax)`, `(x, y, Δy)`, and `errorbar` could have the signatures `(xmin..xmax, ymin..ymax)` and `(x, y, Δx, Δy)`. Would it make sense to merge `errorbarx` and `errorbary` by adding an `orientation` parameter?

**Comment by mkborregaard** [comment]
> Should they be equally spaced, even if it is a numeric axis?

No definitely not IMHO, it's really useful at times to have real numbers. I think either passing it as a CategoricalArray, or even just `string.(`, when you want it to be categories is preferred.

I like the T's but don't feel strongly, but not the midgap plot. The geom_pointinterval is a nice slim look to combine with other plots (series) but not as a standalone imho.

**Comment by sethaxen** [comment]
> The geom_pointinterval is a nice slim look to combine with other plots (series) but not as a standalone imho.

In general I agree, and I'd like to add it as a recipe so it can be combined with other plots. I can think of one use case for it as a standalone. Namely, when getting a quick overview of a posterior sample from a PPL, one usually has many _n_ (many) variables and 4+ chains. A "forest plot" like the one below gives a much more compact (though less informative) overview of the 4 _n_ distributions than e.g. a ridge plot does. You can accomplish the same thing with a boxplot, but IMO when made so thin, the point is more readable than the median line.

43984801-4cc959ee-9cd9-11e8-9cd7-b361ec1d226a
(taken from https://github.com/arviz-devs/arviz/issues/166)

**Comment by sethaxen** [comment]
I fixed the issue with the violin median line extending beyond the edges:
violin_median

Also, I zoomed in on the artifact at the top of the violin plots. Wierdly, the artifact is no longer there when you zoom in, but it reappears when you zoom out:
violin_top

Also, though it's less visible, there's a slight artifact at the bottom that stays when you zoom in:
violin_bottom

**Comment by sethaxen** [comment]
I also added a `trim` option to trim the violin plot to the data limits (using `density`; the default is still untrimmed). It doesn't do any rescaling of the violin width to account for truncation. I personally prefer this as the default but don't care much either way.

```julia
violin(Data(mtcars), :Cyl, :MPG, trim = true)
```
violin_trim

**Comment by piever** [comment]
For trimming I suggest to keep it consistent with `density` (no trim by default). If we change our mind, we can change both in a later PR.

I agree that a mechanism for changing from numeric to categorical should be added somewhere, but probably not here.

The only decision seems to be about the error bar syntax. To the best of my knowledge, related signatures are:

- `shadedline(x, y, ylower, yupper)`
- `ribbon(x, y, yerr)`, same as `shadedline(x, y, y - yerr, y + yerr)`
- `band(x, ylower, yupper)`
- `errorbary(x, y, yerr)`
- `errorbarx(x, y, xerr)`
- `errorbarxy(x, y, xerr, yerr)`

and with this PR also:

- `errorbary(x, y, ymin, ymax)` (does not need `y`)
- `errorbarx(x, y, xmin, xmax)` (does not need `x`)
- `errorbarxy(x, y, xmin xmax, ymin, ymax)`

This is a bit confusing. I would actually suggest to remove all the `y, yerr` and company versions in favor of  `y ± err`.

So a cleaner approach could be:

- `shadedline(x, y, ylower..yupper)`
- `shadedline(x, y ± err)` (`ribbon` recipe goes away)
- `band(x, ylower..yupper)`
- `errorbar(x, ymin..ymax)` (vertical)
- `errorbar(xmin..xmas, y)` (horizontal)

But maybe this kind of revamp can be left for a future PR. Do you need to introduce new signatures for `errorbar` here or is it OK to do that discussion in a separate dedicated PR?

**Comment by mkborregaard** [comment]
@piever interval approach seems not to encompass the (most common) case where the error is supplied as a vector rather than a scalar (error is different for different data points), right?

**Comment by piever** [comment]
> @piever interval approach seems not to encompass the (most common) case where the error is supplied as a vector rather than a scalar (error is different for different data points), right?

Well, one can cheat a bit in that `xmin..xmax` gives an interval with extrema two vectors, but I think we can accept that, otherwise one would need `@. xmin..xmax` to get the vector of intervals (or `y .± err`).

**Comment by sethaxen** [comment]
> So a cleaner approach could be:
> 
> * `shadedline(x, y, ylower..yupper)`
> * `shadedline(x, y ± err)` (`ribbon` recipe goes away)
> * `band(x, ylower..yupper)`
> * `errorbar(x, ymin..ymax)` (vertical)
> * `errorbar(xmin..xmas, y)` (horizontal)

I think this is cleaner as a convenient supported signature with `convert_arguments`, but I'd prefer the signatures to the official plot recipe to still be the expanded ones for explicitness. An `IntervalBased` conversion trait can handle the necessary unpacking, but that would probably go in AbstractPlotting.

> But maybe this kind of revamp can be left for a future PR. Do you need to introduce new signatures for `errorbar` here or is it OK to do that discussion in a separate dedicated PR?

No, we can do that in a future PR.

> Well, one can cheat a bit in that `xmin..xmax` gives an interval with extrema two vectors, but I think we can accept that, otherwise one would need `@. xmin..xmax` to get the vector of intervals (or `y .± err`).

I think the vector extrema is fine, since internally we'll just unpack the vectors anyways. There's also the case where one end of the range is a constant and the other is a vector. Less common, but could happen.

I went ahead and removed T's for the boxplot:

```julia
boxplot(Data(mtcars), :Cyl, :MPG; markersize=10px)
```
boxplot_no_t

And I think that's everything intended for this PR. There are still the open question of the artifact in violin plots, but that can be addressed elsewhere.

File: src/recipes/boxplot.jl [hunk 1]
```diff
@@ -12,20 +10,33 @@ The StatPlots.jl package is licensed under the MIT "Expat" License:
...
+        # notch
+        notch = false,
+        notchwidth = 0.5,
+        # median line
         show_median = true,
+        mediancolor = automatic,
+        medianlinewidth = 1.0,
+        # whiskers
+        range = 1.5, # multiple of IQR controlling whisker length
+        whisker_width = 0.0, # match or multiple of width
```
Comment (piever) [inline src/recipes/boxplot.jl]: Everything is without underscore, so I think we should remove the underscore here as well, even if it is breaking.

File: src/recipes/boxplot.jl [hunk 2]
```diff
@@ -1,8 +1,6 @@
 using AbstractPlotting: extrema_nan
 
-notch_width(q2, q4, N) = 1.58 * (q4-q2)/sqrt(N)
-
-pair_up(dict, key) = (key => dict[key])
+notch_width(q2, q4, N) = 1.58 * (q4 - q2) / sqrt(N)
```
Comment (piever) [inline src/recipes/boxplot.jl]: Even if it is unexported, I guess it makes sense that this also becomes `notchwidth` like the attribute name below.

**Comment by piever** [comment]
Yes, nice job! Let's discuss errorbar syntax and fix violin artifacts in a separate PR. 

Other than the two minor renaming issues, I have a small comment before merging: could you check out AbstractPlotting master in the Travis script, so that we are sure that tests pass on Travis with both master branches?

Otherwise we can simply wait that AbstractPlotting tags a release, but I'd prefer to avoid merging without green CI.

File: src/recipes/boxplot.jl (hunk 1 above)
Comment (sethaxen) [reply src/recipes/boxplot.jl]: Will do. While I'm at it, I'd like to unify the styles for `show_median` and `notch`. either `median`/`notch`, `show_median`/`show_notch`, `showmedian`/`shownotch`, or `medianvisible`/`notchvisible`. Thoughts?

File: src/recipes/boxplot.jl (hunk 2 above)
Comment (sethaxen) [reply src/recipes/boxplot.jl]: Ah, these actually mean different things. `notch_width` here really means notch height, that is, the distance from the median to where the notch ends. `notch_width` in the attribute name is a multiplier of `width` that determines how wide the box is at the notch. I'll change this function to `notchheight` and document what `notchwidth` means.

File: src/recipes/boxplot.jl (hunk 1 above)
Comment (piever) [reply src/recipes/boxplot.jl]: I would try and be compatible with AbstractPlotting `show_axis` and `show_legend` [here](https://github.com/JuliaPlots/AbstractPlotting.jl/blob/master/src/theming.jl#L35), so I would say `show_median`, `show_notch` and `show_midline`.

**Comment by sethaxen** [comment]
Okay, @piever, tests pass. Should I revert the manifest I checked in?

File: src/recipes/errorbar.jl [hunk 3]
```diff
@@ -21,45 +19,84 @@ and horizontal error bars (along the `x`-axis)
...
+    ww = pop!(t, :whiskerwidth)
+    segments = lift(plot[1], plot[2], plot[3], ww) do x,y,Δx,ww
+        bar = Pair.(Point2f0.(x .- Δx, y), Point2f0.(x .+ Δx, y))
+        segments = [bar;]
+        if ww != 0
+            hww = ww ./ 2
+            lw = Pair.(
+                    first.(bar) .- Point2f0.(0, hww),
+                    first.(bar) .+ Point2f0.(0, hww),
+                )
```
Comment (piever) [inline src/recipes/errorbar.jl]:
```suggestion
            lw = Pair.(
                first.(bar) .- Point2f0.(0, hww),
                first.(bar) .+ Point2f0.(0, hww),
            )
```

File: src/recipes/errorbar.jl [hunk 4]
```diff
@@ -21,45 +19,84 @@ and horizontal error bars (along the `x`-axis)
...
+    ww = pop!(t, :whiskerwidth)
+    segments = lift(plot[1], plot[2], plot[3], ww) do x,y,Δy,ww
+        bar = Pair.(Point2f0.(x, y .- Δy), Point2f0.(x, y .+ Δy))
+        segments = [bar;]
+        if ww != 0
+            hww = ww ./ 2
+            lw = Pair.(
+                    first.(bar) .- Point2f0.(hww, 0),
+                    first.(bar) .+ Point2f0.(hww, 0),
+                )
```
Comment (piever) [inline src/recipes/errorbar.jl]:
```suggestion
            lw = Pair.(
                first.(bar) .- Point2f0.(hww, 0),
                first.(bar) .+ Point2f0.(hww, 0),
            )
```

**Comment by piever** [comment]
> Should I revert the manifest I checked in?

No, I think it's best to keep it until tests pass using released version of packages.

**Comment by piever** [comment]
Thanks!

**Pull request merged**